File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-1.java
class Solution {

    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        int[] dp = new int[n+1];

        for (int i = 2; i <= n ; i++){
            dp[i] = Math.min(dp[i-1] + cost[i -1], dp[i-2]+ cost[i - 2]);
        }

        return dp[n];
    }


    // private int [] memo;
    // public int minCostClimbingStairs(int[] cost) {
    //     memo = new int[cost.length];
    //     Arrays.fill(memo, -1);
    //     return Math.min(dfs(cost, 0), dfs(cost, 1));
    // }

    // private int dfs(int[] cost, int i){
    //     if (i >= cost.length) {
    //         return 0;
    //     }

    //     if(memo[i] != -1) return memo[i];

    //     return memo[i] =  cost[i] + Math.min(dfs(cost, i+1), dfs(cost,i+2));
    // }
}

File: Data Structures & Algorithms/climbing-stairs/submission-1.js
class Solution {
    /**
     * @param {number} n
     * @return {number}
     */
    climbStairs(n) {
        const memo = {};

        const dfs = (n , memo) => {
            if (n < 0 ) return 0;
            if (n=== 0 || n === 1) return 1;

            if (memo[n]) return memo[n];

            memo[n] = dfs(n -1, memo) + dfs(n - 2, memo);

            return memo[n]
        }

        return dfs(n, memo);
    }
}
